### PR TITLE
Add SSO feature for Publisher application

### PR DESCRIPTION
**Pull request by npamudika**
### Proposed changes in this pull request

- SSO implementation for Publisher application

### When should this PR be merged

ASAP

### Follow up actions

[List any possible follow-up actions here; for instance, testing data
migrations, software that we need to install on staging and production
environments.]

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.authenticator/src/main/java/org/wso2/carbon/apimgt/rest/api/authenticator/AuthenticatorService.java [hunk 1]
```diff
@@ -80,8 +80,9 @@ public JsonObject getAuthenticationConfigurations(String appName)
         grantTypes.add(KeyManagerConstants.PASSWORD_GRANT_TYPE);
         grantTypes.add(KeyManagerConstants.AUTHORIZATION_CODE_GRANT_TYPE);
         grantTypes.add(KeyManagerConstants.REFRESH_GRANT_TYPE);
-        APIMStoreConfigurations storeConfigs = ServiceReferenceHolder.getInstance().getAPIMStoreConfiguration();
-        String callBackURL = storeConfigs.getApimBaseUrl() + "login/callback/" + appName;
+        APIMAppConfigurations appConfigs = ServiceReferenceHolder.getInstance()
+                .getAPIMAppConfiguration();
+        String callBackURL = appConfigs.getApimBaseUrl() + "login/callback/" + appName;
```
Comment (anuruddhal) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.authenticator/src/main/java/org/wso2/carbon/apimgt/rest/api/authenticator/AuthenticatorService.java]: Shall we move "login/callback/" to a constant?

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.authenticator/src/main/java/org/wso2/carbon/apimgt/rest/api/authenticator/AuthenticatorAPI.java [hunk 2]
```diff
@@ -280,26 +282,43 @@ public Response callback(@Context Request request, @PathParam("appName") String
...
+                            + authResponseBean.getIdToken() + "&partial_token=" + authResponseBean.getPartialToken()
+                            + "&scopes=" + authResponseBean.getScopes() + "&validity_period="
+                            + authResponseBean.getValidityPeriod();
+                    URI redirectURI = new URI(appConfigs.getApimBaseUrl() + "publisher/login?user_name="
+                            + URLEncoder.encode(authResponseBeanData, "UTF-8")
+                            .replaceAll("\\+", "%20").replaceAll("%26", "&")
+                            .replaceAll("%3D", "="));
+                    return Response.status(Response.Status.FOUND)
+                            .header(HttpHeaders.LOCATION, redirectURI)
+                            .cookie(cookieWithAppContext, httpOnlyCookieWithAppContext,
```
Comment (prasa7) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.authenticator/src/main/java/org/wso2/carbon/apimgt/rest/api/authenticator/AuthenticatorAPI.java]: formatting issue

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.authenticator/src/main/java/org/wso2/carbon/apimgt/rest/api/authenticator/AuthenticatorAPI.java [hunk 3]
```diff
@@ -280,26 +282,43 @@ public Response callback(@Context Request request, @PathParam("appName") String
...
+                                    restAPIContextCookie)
+                            .build();
+                } else {
+                    return Response.status(Response.Status.FOUND)
+                            .header(HttpHeaders.LOCATION, targetURIForRedirection).entity(authResponseBean)
+                            .cookie(cookieWithAppContext, httpOnlyCookieWithAppContext,
+                                    restAPIContextCookie, authUserCookie)
```
Comment (prasa7) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.authenticator/src/main/java/org/wso2/carbon/apimgt/rest/api/authenticator/AuthenticatorAPI.java]: formatting issue

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.authenticator/src/main/java/org/wso2/carbon/apimgt/rest/api/authenticator/AuthenticatorService.java [hunk 4]
```diff
@@ -80,8 +80,9 @@ public JsonObject getAuthenticationConfigurations(String appName)
         grantTypes.add(KeyManagerConstants.PASSWORD_GRANT_TYPE);
         grantTypes.add(KeyManagerConstants.AUTHORIZATION_CODE_GRANT_TYPE);
         grantTypes.add(KeyManagerConstants.REFRESH_GRANT_TYPE);
-        APIMStoreConfigurations storeConfigs = ServiceReferenceHolder.getInstance().getAPIMStoreConfiguration();
-        String callBackURL = storeConfigs.getApimBaseUrl() + "login/callback/" + appName;
+        APIMAppConfigurations appConfigs = ServiceReferenceHolder.getInstance()
```
Comment (prasa7) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.authenticator/src/main/java/org/wso2/carbon/apimgt/rest/api/authenticator/AuthenticatorService.java]: Check whether we can get to a single line as 
ServiceReferenceHolder.getInstance()..getAPIMAppConfiguration();

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.webserver/src/main/java/org/wso2/carbon/apimgt/rest/api/webserver/impl/PublisherApiServiceImpl.java [hunk 5]
```diff
@@ -48,7 +48,7 @@
             //#TODO read from config file
         /* TODO: Check the dot containment in last segment separated by '/' or use regex to capture file extension */
             if (app.equals(context)) {
-                if (rawUri.contains(".")) {
+                if (rawUri.split("\\?")[0].contains(".")) {
                     filePath = "./deployment/webapps" + path;
```
Comment (anuruddhal) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.webserver/src/main/java/org/wso2/carbon/apimgt/rest/api/webserver/impl/PublisherApiServiceImpl.java]: Move "./deployment/webapps" to a string constant.
Can't we remove the TODO comment now?

File: features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/source/src/App.js [hunk 6]
```diff
@@ -68,11 +76,22 @@ class Protected extends Component {
...
+                var scopes = this.state.authConfigs.scopes;
+                window.location = authorizationEndpoint+"?response_type=code&client_id="+client_id+"&redirect_uri="+callback_URL+"&scope="+scopes;
+            } else {
+                return (
+                            <Redirect to={{pathname: '/login', search: params}}/>
+                        );
+            }
+        } else {
+            return null;
+        }
```
Comment (isurulucky) [inline features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/source/src/App.js]: returning null will stop the component from rendering AFAIK. Should display an error to the user IMHO.

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.authenticator/src/main/java/org/wso2/carbon/apimgt/rest/api/authenticator/AuthenticatorAPI.java [hunk 7]
```diff
@@ -280,26 +282,43 @@ public Response callback(@Context Request request, @PathParam("appName") String
...
+                URI targetURIForRedirection = new URI(appConfigs.getApimBaseUrl() + appName);
+                if (AuthenticatorConstants.PUBLISHER_APPLICATION.equals(appName)) {
+                    String authResponseBeanData = authResponseBean.getAuthUser() + "&id_token="
+                            + authResponseBean.getIdToken() + "&partial_token=" + authResponseBean.getPartialToken()
+                            + "&scopes=" + authResponseBean.getScopes() + "&validity_period="
+                            + authResponseBean.getValidityPeriod();
+                    URI redirectURI = new URI(appConfigs.getApimBaseUrl() + "publisher/login?user_name="
+                            + URLEncoder.encode(authResponseBeanData, "UTF-8")
+                            .replaceAll("\\+", "%20").replaceAll("%26", "&")
+                            .replaceAll("%3D", "="));
```
Comment (isurulucky) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.authenticator/src/main/java/org/wso2/carbon/apimgt/rest/api/authenticator/AuthenticatorAPI.java]: Can you please explain why its necessary to use these replaceAll code segments?

Comment (npamudika) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.authenticator/src/main/java/org/wso2/carbon/apimgt/rest/api/authenticator/AuthenticatorAPI.java]: We have to handle some of the characters separately as java.net.URLEncoder does not support them.

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.webserver/src/main/java/org/wso2/carbon/apimgt/rest/api/webserver/impl/PublisherApiServiceImpl.java (hunk 5 above)
Comment (npamudika) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.webserver/src/main/java/org/wso2/carbon/apimgt/rest/api/webserver/impl/PublisherApiServiceImpl.java]: @tmkasun has addressed this.

File: features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/source/src/App.js [hunk 8]
```diff
@@ -35,10 +37,17 @@ import './App.css'
...
     constructor(props) {
         super(props);
-        this.state = {showLeftMenu: false};
+        this.state = {showLeftMenu: false, authConfigs: null};
         this.setLeftMenu = this.setLeftMenu.bind(this);
         message.config({top: '48px'}); // .custom-header height + some offset
         /* TODO: need to fix the header to avoid conflicting with messages ~tmkb*/
+        this.handleResponse = this.handleResponse.bind(this);
+        /* TODO: Get apim base url from config*/
+        Axios.get("https://localhost:9292/login/login/publisher").then(this.handleResponse);
```
Comment (dakshika) [inline features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/source/src/App.js]: Do not hardcode the URL

Comment (npamudika) [reply features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/source/src/App.js]: We have already created a git issue for this and will be tracked via https://github.com/wso2/carbon-apimgt/issues/5087.

**Pull request merged**